packages/devtools_app/lib/src/screens/memory/panes/chart/chart_pane.dart

```diff
@@ -0,0 +1,706 @@
+
+class MemoryChartPane extends StatefulWidget {
+  const MemoryChartPane({
+    Key? key,
+    required this.chartControllers,
+    required this.keyPressed,
+  }) : super(key: key);
+  final ChartControllers chartControllers;
+  final ValueListenable<RawKeyEvent?> keyPressed;
+  static final hoverKey = GlobalKey(debugLabel: 'Chart Hover');
+
+  @override
+  State<MemoryChartPane> createState() => _MemoryChartPaneState();
+}
+
+class _MemoryChartPaneState extends State<MemoryChartPane>
+    with
+        AutoDisposeMixin,
+        SingleTickerProviderStateMixin,
+        ProvidedControllerMixin<MemoryController, MemoryChartPane> {
+  OverlayEntry? _hoverOverlayEntry;
+
+  static const _hoverXOffset = 10;
+  static const _hoverYOffset = 0.0;
+  static double get _hoverWidth => scaleByFontFactor(225.0);
+  static const _hover_card_border_width = 2.0;
+
+  // TODO(terry): Compute below heights dynamically.
+  static double get _hoverHeightMinimum => scaleByFontFactor(42.0);
+  static double get hoverItemHeight => scaleByFontFactor(18.0);
+
+  // One extension event to display (4 lines).
+  static double get _hoverOneEventsHeight => scaleByFontFactor(82.0);
+
+  // Many extension events to display.
+  static double get _hoverEventsHeight => scaleByFontFactor(120.0);
+
+  static double _computeHoverHeight(
+    int eventsCount,
+    int tracesCount,
+    int extensionEventsCount,
+  ) =>
+      _hoverHeightMinimum +
+      (eventsCount * hoverItemHeight) +
+      _hover_card_border_width +
+      (tracesCount * hoverItemHeight) +
+      (extensionEventsCount > 0
+          ? (extensionEventsCount == 1
+              ? _hoverOneEventsHeight
+              : _hoverEventsHeight)
+          : 0);
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    if (!initController()) return;
+
+    addAutoDisposeListener(widget.keyPressed, () {
+      final event = widget.keyPressed.value;
+      if (event == null) return;
+      if (event.isKeyPressed(LogicalKeyboardKey.escape)) {
+        _hideHover();
+      }
+    });
+
+    addAutoDisposeListener(widget.chartControllers.event.tapLocation, () {
+      if (widget.chartControllers.event.tapLocation.value != null) {
+        if (_hoverOverlayEntry != null) {
+          _hideHover();
+        }
+        final tapLocation = widget.chartControllers.event.tapLocation.value;
+        if (tapLocation?.tapDownDetails != null) {
+          final tapData = tapLocation!;
+          final index = tapData.index;
+          final timestamp = tapData.timestamp!;
+
+          final copied = TapLocation.copy(tapLocation);
+          widget.chartControllers.vm.tapLocation.value = copied;
+          widget.chartControllers.android.tapLocation.value = copied;
+
+          final allValues = ChartsValues(controller, index, timestamp);
+          _showHover(
+            context,
+            allValues,
+            tapData.tapDownDetails!.globalPosition,
+          );
+        }
+      }
```

Comment:
Can you either put all this in a helper or add a TODO to do so? The code for all three of these listeners is almost identical and can be generalized a bit to be put in a helper method.